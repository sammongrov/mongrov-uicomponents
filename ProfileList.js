import React, { Component } from 'react';
import { FlatList, View, Text } from 'react-native';
import { iOSColors } from 'react-native-typography';
import { ListItem } from 'react-native-elements';
import PropTypes from 'prop-types';
import { styles } from 'react-native-theme';
import { Icon, Avatar } from '@ui/components';
import { Colors } from '@ui/theme_default';

export default class ProfileList extends Component {
  static propTypes = {};

  static defaultProps = {};

  keyExtractor = (item) => item.id;

  renderItem = ({ item }) => {
    const { profileContainerStyle } = this.props;
    let statusColor = iOSColors.gray;
    switch (item.status) {
      case 'online':
        statusColor = iOSColors.green;
        break;
      case 'away':
        statusColor = iOSColors.yellow;
        break;
      case 'busy':
        statusColor = iOSColors.red;
        break;
      default:
        statusColor = iOSColors.midGray;
    }
    return (
      <ListItem
        key={item.id}
        title={item.title}
        subtitle={item.subTitle ? item.subTitle : null}
        leftIcon={
          item.icon && (
            <Icon
              name={item.icon}
              type="material-community"
              color={item.title === 'Logout' ? Colors.TEXT_HEADER : iOSColors.gray}
              size={26}
            />
          )
        }
        leftAvatar={
          item.avatar_url &&
          statusColor && (
            <Avatar
              avatarUrl={item.avatar_url}
              avatarName={item.title}
              avatarSize={50}
              statusColor={statusColor}
              onAvatarPress={item.onAvatarPress ? item.onAvatarPress : null}
            />
          )
        }
        containerStyle={[styles.listItemContainerStyle, profileContainerStyle]}
        onPress={item.onPress ? item.onPress : null}
        titleStyle={
          item.subTitle
            ? styles.listItemTitle
            : [
                styles.fontSize16,

                {
                  color: item.title === 'Logout' ? Colors.TEXT_HEADER : iOSColors.black,
                },
              ]
        }
        subtitleStyle={styles.listItemSubTitle}
        rightTitle={
          item.date &&
          item.time && (
            <View style={{ flexDirection: 'column' }}>
              <Text style={{ color: iOSColors.gray, fontSize: 14, textAlign: 'right' }}>
                {item.date ? item.date : ''}
              </Text>
              <Text style={{ color: iOSColors.gray, fontSize: 12, textAlign: 'right' }}>
                {item.time ? item.time : ''}
              </Text>
            </View>
          )
        }
      />
    );
  };

  render() {
    const { list } = this.props;
    return <FlatList keyExtractor={this.keyExtractor} data={list} renderItem={this.renderItem} />;
  }
}

ProfileList.propTypes = {
  profileContainerStyle: PropTypes.shape(),
  list: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      icon: PropTypes.string,
      onPress: PropTypes.func,
    }),
  ),
};

ProfileList.defaultProps = {
  profileContainerStyle: null,
  list: PropTypes.arrayOf(
    PropTypes.shape({
      id: '',
      title: '',
      icon: '',
      onPress: () => {},
    }),
  ),
};
