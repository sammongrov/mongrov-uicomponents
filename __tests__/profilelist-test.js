import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Alert } from 'react-native';

import ProfileList from '../ProfileList';

configure({ adapter: new Adapter() });
const onPress = () => Alert.alert('onPress');

it('ProfileList renders correctly', () => {
  const list = [
    {
      id: '0',
      title: 'smith@gmail.com',
      icon: 'email-outline',
      status: 'online',
      onPress: jest.fn(),
    },
    {
      id: '1',
      title: 'Link Facebook account',
      icon: 'facebook-box',
      status: 'online',
      onPress: jest.fn(),
    },
    {
      id: '2',
      title: 'Link Google account',
      icon: 'google',
      avatar_url: 'https://my-avatar.com/avatar-215.png',
      status: 'busy',
      onPress: jest.fn(),
      onAvatarPress: jest.fn(),
    },
    {
      id: '3',
      title: 'Link phone number',
      icon: 'phone',
      onPress: jest.fn(),
    },
    {
      id: '4',
      title: 'Logout',
      icon: 'logout',
      onPress: jest.fn(),
      style: 'color:2E88FF',
      date: '05.02.19',
      time: '12:16PM',
    },
  ];
  const tree = renderer.create(<ProfileList onPress={onPress} list={list} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('onAvatarPress is called', () => {
  const item = {
    id: '2',
    title: 'Link Google account',
    icon: 'google',
    avatar_url: 'https://my-avatar.com/avatar-215.png',
    status: 'busy',
    onAvatarPress: jest.fn(),
  };
  const root = shallow(<ProfileList list={[item]} />);
  const rootInstance = root.instance();
  const avatar = shallow(rootInstance.renderItem({ item })).find('Avatar');
  avatar.props().onAvatarPress();
  expect(item.onAvatarPress.mock.calls.length).toBe(1);
});
