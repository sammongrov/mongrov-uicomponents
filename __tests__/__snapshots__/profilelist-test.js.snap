// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileList renders correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "icon": "email-outline",
        "id": "0",
        "onPress": [MockFunction],
        "status": "online",
        "title": "smith@gmail.com",
      },
      Object {
        "icon": "facebook-box",
        "id": "1",
        "onPress": [MockFunction],
        "status": "online",
        "title": "Link Facebook account",
      },
      Object {
        "avatar_url": "https://my-avatar.com/avatar-215.png",
        "icon": "google",
        "id": "2",
        "onAvatarPress": [MockFunction],
        "onPress": [MockFunction],
        "status": "busy",
        "title": "Link Google account",
      },
      Object {
        "icon": "phone",
        "id": "3",
        "onPress": [MockFunction],
        "title": "Link phone number",
      },
      Object {
        "date": "05.02.19",
        "icon": "logout",
        "id": "4",
        "onPress": [MockFunction],
        "style": "color:2E88FF",
        "time": "12:16PM",
        "title": "Logout",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flexDirection": "row",
                "padding": 14,
              },
              undefined,
              Array [
                undefined,
                null,
              ],
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#8E8E93",
                  "fontSize": 26,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="material-community"
          >
            
          </Text>
          <View
            width={16}
          />
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "fontSize": 17,
                      },
                      Array [
                        undefined,
                        Object {
                          "color": "#000000",
                        },
                      ],
                    ],
                    undefined,
                  ],
                ]
              }
            >
              smith@gmail.com
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flexDirection": "row",
                "padding": 14,
              },
              undefined,
              Array [
                undefined,
                null,
              ],
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#8E8E93",
                  "fontSize": 26,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="material-community"
          >
            
          </Text>
          <View
            width={16}
          />
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "fontSize": 17,
                      },
                      Array [
                        undefined,
                        Object {
                          "color": "#000000",
                        },
                      ],
                    ],
                    undefined,
                  ],
                ]
              }
            >
              Link Facebook account
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flexDirection": "row",
                "padding": 14,
              },
              undefined,
              Array [
                undefined,
                null,
              ],
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#8E8E93",
                  "fontSize": 26,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="material-community"
          >
            
          </Text>
          <View
            width={16}
          />
          <View
            style={
              Array [
                undefined,
                Object {
                  "position": "relative",
                },
              ]
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#6c1f34",
                  "borderRadius": 25,
                  "height": 50,
                  "opacity": 1,
                  "width": 50,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 25,
                      "height": 50,
                      "width": 50,
                    },
                  ]
                }
              >
                <FastImageView
                  onFastImageError={[Function]}
                  resizeMode="cover"
                  source={
                    Object {
                      "priority": "high",
                      "uri": "https://my-avatar.com/avatar-215.png",
                    }
                  }
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                Array [
                  undefined,
                  Object {
                    "backgroundColor": "#FF3B30",
                    "position": "absolute",
                    "right": 0,
                  },
                ]
              }
            />
          </View>
          <View
            width={16}
          />
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "fontSize": 17,
                      },
                      Array [
                        undefined,
                        Object {
                          "color": "#000000",
                        },
                      ],
                    ],
                    undefined,
                  ],
                ]
              }
            >
              Link Google account
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flexDirection": "row",
                "padding": 14,
              },
              undefined,
              Array [
                undefined,
                null,
              ],
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#8E8E93",
                  "fontSize": 26,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="material-community"
          >
            
          </Text>
          <View
            width={16}
          />
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "fontSize": 17,
                      },
                      Array [
                        undefined,
                        Object {
                          "color": "#000000",
                        },
                      ],
                    ],
                    undefined,
                  ],
                ]
              }
            >
              Link phone number
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flexDirection": "row",
                "padding": 14,
              },
              undefined,
              Array [
                undefined,
                null,
              ],
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#2E88FF",
                  "fontSize": 26,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="material-community"
          >
            
          </Text>
          <View
            width={16}
          />
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "fontSize": 17,
                      },
                      Array [
                        undefined,
                        Object {
                          "color": "#2E88FF",
                        },
                      ],
                    ],
                    undefined,
                  ],
                ]
              }
            >
              Logout
            </Text>
          </View>
          <View
            width={16}
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "flex-end",
                  "flex": 0.5,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#8E8E93",
                    "fontSize": 14,
                    "textAlign": "right",
                  }
                }
              >
                05.02.19
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#8E8E93",
                    "fontSize": 12,
                    "textAlign": "right",
                  }
                }
              >
                12:16PM
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
